grand_challenge_forge/generation_utils.py

```diff
@@ -59,3 +63,59 @@ def ci_to_civ(component_interface):
         **civ,
         "interface": component_interface,
     }
+
+
+def get_jinja2_environment(searchpath=None):
+    from grand_challenge_forge.partials.filters import custom_filters
+
+    env = ImmutableSandboxedEnvironment(
+        loader=FileSystemLoader(searchpath=searchpath or []),
+        undefined=StrictUndefined,
+    )
+    env.filters = custom_filters
+    env.globals["now"] = datetime.now(timezone.utc)
+
+    return env
+
+
+def copy_and_render(
+    *,
+    source_path,
+    output_path,
+    context,
+):
+    # Create the output directory if it doesn't exist
+    output_path.mkdir(parents=True, exist_ok=True)
+
+    env = get_jinja2_environment(searchpath=source_path)
+
+    for root, _, files in os.walk(source_path):
+        root = Path(root)
+        # Create relative path
+        rel_path = root.relative_to(source_path)
+        current_output_dir = output_path / rel_path
+
+        # Create directories in the output path
+        current_output_dir.mkdir(parents=True, exist_ok=True)
+
+        for file in files:
```

Comment:
Could we check that we're templating files from the intended directory for safety?

```python
if root.resolve() not in file.resolve().parents:
  raise RuntimeError(...)
```

We could make this more strict and ensure that all templates are coming from `__file__ / "partials"`.

Reply:
Good, nefarious thoughts! I'll have a look.

Reply:
Continued down the path of nefarious scenarios and added tweaked things with safety in mind, albeit it requires an actor to control the filesystem. Having said that, these tweaks do reduce chances of the code being use at the wrong spot.

- The calling of the function now involves a template name instead of a Path, core assumption is made that it then should be findable in the template directories.
- File and directory soft links are only allowed if the resolved path is within the template directories.
- Templates can include other templates either directly from local directory, or via a path resolvable relative to the "partials"

I think that should lock down things enough.

Reply:
Ofc, added some tests to cover these scenarios.